subsys/net/lib/coap/coap_client.c

```diff
@@ -811,10 +811,19 @@ static int handle_response(struct coap_client *client, const struct coap_packet
 	}
 fail:
 	client->response_ready = false;
-	internal_req->request_ongoing = false;
+	if (ret != 0 || !coap_request_is_observe(&internal_req->request)) {
+		internal_req->request_ongoing = false;
+	}
```

Comment:
I initially commented about this here:
https://github.com/zephyrproject-rtos/zephyr/pull/72540#pullrequestreview-2061630427

But I realize it might make more sense to discuss on this PR

It seems to me this diff ***might*** have introduced an unintended bug? @tautologyclub @juhaylinen can you confirm whether this is true?

IE, did you really mean `ret != 0` here, or should it be `ret < 0`?

Here is my thinking (copied from that original comment):

The only two functions that `handle_response` calls that can return positive return values are `send_request` and `send_ack`, both of which use positive values to indicate the number of bytes sent.

But for `send_request`, this gets converted to an early `return 1`, whereas for `send_ack`, this value ***might*** (under the right conditions) be returned directly via the `fail` block?

Presumably, the difference here is that `send_request` results in an as-yet ongoing request, whereas `send_ack` completes the request?

So, originally, the `fail` block was essentially handling all scenarios where the request finishes -- either due to error, or simply having completed successfully. 

*This is where we get to the bug*:

What if:
1. `send_ack` succeeds
2. `block_option` happens to be non-positive
3. `blockwise_transfer` happens to be false, or `last_block` happens to be true
4. The request is of OBSERVE type

If those four conditions are met, then the request will be incorrectly marked as no-longer ongoing, won't it? Or is this just not possible? Or is it intended, for some reason I don't understand?

Reply:
I think probably there are better places for this discussion still :D 

You are correct I think. I have some local fixes to this PR which I haven't pushed because I was just kinda baffled by the insta-merge :D I'll create a new PR - hopefully I have time today (btw I'm @luegofuego on Discord if you wanna discuss more in depth)

Reply:
It should definitely be `ret < 0` but keep in mind we also have the problem of the request possibly not being in scope anymore plus the requirement of MID-based deduplication. I have code for all this though.

Reply:
See https://github.com/zephyrproject-rtos/zephyr/pull/72948